README.rst
===============================
instant2fa
===============================

Instant2FA Python bindings.

Installation
------------

Install it via PyPI:

.. code-block:: shell

    pip install instant2fa

Usage
-----

.. code-block:: python

    import instant2fa

    instant2fa.access_key = 'YOUR_ACCESS_KEY'
    instant2fa.access_secret = 'YOUR_ACCESS_SECRET'

    distinct_id = 'A_UNIQUE_ID_FOR_A_USER'

    hosted_page_url = instant2fa.create_settings(distinct_id)
    hosted_page_url = instant2fa.create_verification(distinct_id)
    verification_succeeded = instant2fa.confirm_verification(distinct_id, token)

Development
-----------

To start development, run:

.. code-block:: shell

    $ git clone git@github.com:clef/instant2fa-python.git instant2fa
    $ cd instant2fa
    $ make develop


HISTORY.rst
=======
History
=======

1.0.0
------------------

* First release on PyPI.

1.0.1
-----------------

* Validate `distinct_id` parameter has a value.


tests/test_instant2fa.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
test_instant2fa
----------------------------------

Tests for `instant2fa` module.
"""
import pytest
import requests
import requests_mock

import instant2fa
from instant2fa import errors

TOKEN_ID = 'tok_123'
DISTINCT_ID = 'A_UNIQUE_ID'

USER_ACCESS_TOKEN_URL = '{}/user-access-tokens/'.format(instant2fa.api_base)
VERIFICATION_REQUEST_URL = '{}/verification-requests/'.format(
    instant2fa.api_base
)
VERIFICATION_RESPONSE_TOKEN_URL = '{}/verification-response-tokens/tok_123'.format(
    instant2fa.api_base
)


@pytest.fixture
def user_access_token():
    return {
        'data': {
            'attributes': {
                'hosted_page_url': 'test_hosted_pages',
                'distinct_id': DISTINCT_ID
            },
            'type': 'user-access-tokens'
        }
    }


@pytest.fixture
def verification_request():
    return {
        'data': {
            'attributes': {
                'hosted_page_url': 'test_hosted_pages',
                'distinct_id': DISTINCT_ID
            },
            'type': 'verification-requests'
        }
    }


@pytest.fixture
def verification_response_token():
    return {
        'data': {
            'attributes': {
                'status': 'succeeded',
                'distinct_id': DISTINCT_ID
            },
            'type': 'verification-responses'
        }
    }


@pytest.fixture
def failed_verification_response_token():
    return {
        'data': {
            'attributes': {
                'status': 'failed',
                'distinct_id': DISTINCT_ID
            },
            'type': 'verification-responses'
        }
    }


@pytest.fixture
def error():
    return {
        'errors': [{
            "status": 400,
            "detail": "Something went wrong.",
            "title": "ValidationException"
        }]
    }


@pytest.fixture
def configured_instant2fa():
    instant2fa.access_key = 'ACCESS_KEY'
    instant2fa.access_secret = 'ACCESS_SECRET'
    return instant2fa


def test_authentication():
    instant2fa.access_key = None
    instant2fa.access_secret = None
    with pytest.raises(errors.AuthenticationError):
        instant2fa.create_settings(DISTINCT_ID)


def test_create_settings(configured_instant2fa, user_access_token):

    with requests_mock.Mocker() as mock:

        mock.request(
            'POST',
            USER_ACCESS_TOKEN_URL,
            json=user_access_token,
            status_code=201
        )
        hosted_pages_url = configured_instant2fa.create_settings(DISTINCT_ID)
        assert hosted_pages_url == 'test_hosted_pages'


def test_create_settings_error(configured_instant2fa, error):
    with requests_mock.Mocker() as mock:
        mock.request('POST', USER_ACCESS_TOKEN_URL, json=error, status_code=401)
        with pytest.raises(errors.APIError):
            configured_instant2fa.create_settings(DISTINCT_ID)


def test_create_verification(configured_instant2fa, verification_request):
    with requests_mock.Mocker() as mock:
        mock.request(
            'POST',
            VERIFICATION_REQUEST_URL,
            json=verification_request,
            status_code=201
        )
        hosted_pages_url = configured_instant2fa.create_verification(
            DISTINCT_ID
        )
        assert hosted_pages_url == 'test_hosted_pages'


def test_create_verification_for_non_mfa_enabled_user(
    configured_instant2fa, verification_request
):
    with requests_mock.Mocker() as mock:
        mock.request(
            'POST',
            VERIFICATION_REQUEST_URL,
            json=verification_request,
            status_code=422
        )
        with pytest.raises(errors.MFANotEnabled):
            configured_instant2fa.create_verification(DISTINCT_ID)


def test_confirm_verification(
    configured_instant2fa, verification_response_token
):
    with requests_mock.Mocker() as mock:
        mock.request(
            'GET',
            VERIFICATION_RESPONSE_TOKEN_URL,
            json=verification_response_token,
            status_code=202
        )
        confirmed = configured_instant2fa.confirm_verification(
            DISTINCT_ID, TOKEN_ID
        )
        assert confirmed


def test_confirm_verification_mismatch(
    configured_instant2fa, verification_response_token
):
    with requests_mock.Mocker() as mock:
        mock.request(
            'GET',
            VERIFICATION_RESPONSE_TOKEN_URL,
            json=verification_response_token,
            status_code=202
        )
        with pytest.raises(errors.VerificationMismatch):
            configured_instant2fa.confirm_verification(
                'mismatched_id', TOKEN_ID
            )


def test_confirm_verification_failed(
    configured_instant2fa, failed_verification_response_token
):
    with requests_mock.Mocker() as mock:
        mock.request(
            'GET',
            VERIFICATION_RESPONSE_TOKEN_URL,
            json=failed_verification_response_token,
            status_code=202
        )
        with pytest.raises(errors.VerificationFailed):
            configured_instant2fa.confirm_verification(DISTINCT_ID, TOKEN_ID)


instant2fa/jsonapi.py
def get_attributes_from_response(response):
    data = response.json().get('data')
    return data.get('attributes')


def construct_path(resource_type, method, lookup_key=None):
    if method == 'POST':
        return '/{}/'.format(resource_type)
    elif method == 'GET':
        return '/{}/{}'.format(resource_type, lookup_key)
    else:
        raise NotImplementedError('Method {} not supported.'.format(method))


def construct_request_body(resource_type, **kwargs):
    top_level = dict()
    data = dict(type=resource_type)
    attributes = dict()
    for k, v in kwargs.items():
        attributes[k] = v
    data['attributes'] = attributes
    top_level['data'] = data
    return top_level


def get_error_from_response(response):
    error = response.json().get('errors').pop()
    message = error.get('detail') or error.get('title')
    return message


def get_headers():
    return {'Content-Type': 'application/vnd.api+json'}


instant2fa/api_requestor.py
import json
from urlparse import urljoin

import requests

import instant2fa
from instant2fa import errors
from instant2fa import jsonapi

class APIRequestor(object):

    def __init__(
        self, access_key=None, access_secret=None, api_base=None, headers=None
    ):
        self.api_base = api_base or instant2fa.api_base
        self.access_key = access_key or instant2fa.access_key
        self.access_secret = access_secret or instant2fa.access_secret
        self.headers = headers or jsonapi.get_headers()

    def request(self, method, path, body=None, accepted_codes=None):
        if not self.access_key:
            raise errors.AuthenticationError(
                'No access key provided. (HINT: set your access key using '
                '"instant2fa.access_key = <ACCESS_KEY>"). You can generate '
                'access keys from the Instant2FA dashboard. '
                'See https://dashboard.instant2fa.com/login'
            )

        if not self.access_secret:
            raise errors.AuthenticationError(
                'No access secret provided. (HINT: set your access secret '
                'using "instant2fa.access_secret = <ACCESS_SECRET>"). You can '
                'generate  access secrets from the Instant2FA dashboard. '
                'See https://dashboard.instant2fa.com/login'
            )
        accepted_codes = accepted_codes or []
        uri = urljoin(self.api_base, path)
        auth = (self.access_key, self.access_secret)
        response = requests.request(
            method,
            uri,
            data=json.dumps(body),
            auth=auth,
            headers=self.headers
        )
        if response.status_code in accepted_codes:
            return response
        else:
            return self.handle_api_error(response)

    def handle_api_error(self, response):
        try:
            response.json()
        except ValueError:
            raise errors.APIError(
                "Invalid response object from API. HTTP response code "
                "was {}".format(response.status_code)
            )
        raise errors.APIError(jsonapi.get_error_from_response(response))


instant2fa/resources.py
from instant2fa import errors
from instant2fa.api_requestor import APIRequestor
from instant2fa import jsonapi


class Instant2FAResource(object):

    def __init__(self, **kwargs):
        for k, v in kwargs.items():
            setattr(self, k, v)


class UserAccessToken(Instant2FAResource):
    resource_type = 'user-access-tokens'

    @classmethod
    def create(cls, distinct_id, method='POST'):
        body = jsonapi.construct_request_body(
            cls.resource_type, distinct_id=distinct_id
        )
        path = jsonapi.construct_path(cls.resource_type, method)
        response = APIRequestor().request(
            method, path, body=body, accepted_codes=[201]
        )
        return cls(**jsonapi.get_attributes_from_response(response))


class VerificationRequest(Instant2FAResource):
    resource_type = 'verification-requests'

    @classmethod
    def handle_response(cls, response):
        if response.status_code == 422:
            raise errors.MFANotEnabled
        return cls(**jsonapi.get_attributes_from_response(response))

    @classmethod
    def create(cls, distinct_id, method='POST'):
        body = jsonapi.construct_request_body(
            cls.resource_type, distinct_id=distinct_id
        )
        path = jsonapi.construct_path(cls.resource_type, method)
        response = APIRequestor().request(
            method, path, body=body, accepted_codes=[201, 422]
        )
        return cls.handle_response(response)


class VerificationResponseToken(Instant2FAResource):
    resource_type = 'verification-response-tokens'

    @classmethod
    def retrieve(cls, token, method='GET'):
        path = jsonapi.construct_path(
            cls.resource_type, method, lookup_key=token
        )
        response = APIRequestor().request(method, path, accepted_codes=[202])
        return cls(**jsonapi.get_attributes_from_response(response))


def create_settings(distinct_id):
    """
    Args:
        distinct_id: string
    Returns:
        string: Hosted settings page URL

    """
    if not distinct_id:
        raise errors.ValidationError('A value is required for distinct_id.')
    user_access_token = UserAccessToken.create(str(distinct_id))
    return user_access_token.hosted_page_url


def create_verification(distinct_id):
    """
    Args:
        distinct_id: string
    Returns:
        string: Hosted verification page URL
    """
    if not distinct_id:
        raise errors.ValidationError('A value is required for distinct_id.')
    verification = VerificationRequest.create(str(distinct_id))
    return verification.hosted_page_url


def confirm_verification(distinct_id, token):
    """
    Args:
        distinct_id: string
        token: string
    Returns:
        boolean
    Raises:
        VerificationMismatch
        VerificationFailed
    """
    if not distinct_id or not token:
        raise errors.ValidationError(
            'A value is required for distinct_id and token params.'
        )
    verification_response = VerificationResponseToken.retrieve(token)
    if verification_response.distinct_id != str(distinct_id):
        raise errors.VerificationMismatch(
            "The distinct_id passed back from the request didn't match the "
            "one passed into the function. Are you passing in the right value for "
            "distinct_id?"
        )
    if verification_response.status != 'succeeded':
        raise errors.VerificationFailed(
            "The verification did not pass. The status was {}".
            format(verification_response.status)
        )
    return True
